### PR TITLE
CONTRIB-8674: Write additional unit testing

### PR DESCRIPTION
**Pull request by laurentdavid**
* Backup/restore tests
* Use a trait instead of a superclass for tests

### CODE REVIEW

File: classes/test/testcase_helper_trait.php [hunk 1]
```diff
@@ -16,19 +16,21 @@
 
 namespace mod_bigbluebuttonbn\test;
 
-use advanced_testcase;
 use context_module;
+use PHPUnit\Framework\SkippedTestError;
```
Comment (andrewnicols) [inline classes/test/testcase_helper_trait.php:20]: Why are these necessary?

File: classes/test/testcase_helper_trait.php [hunk 2]
```diff
@@ -148,4 +158,33 @@ protected function require_mock_server(): void {
             );
         }
     }
+    // phpcs:disable moodle.NamingConventions.ValidFunctionName.LowercaseFunction
```
Comment (andrewnicols) [inline classes/test/testcase_helper_trait.php:161]: Why? These are already part of the advanced_testcase

File: tests/backup_restore_test.php [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+     */
+    public function setUp(): void {
+        parent::setUp();
+        $this->basic_setup();
+    }
+
+    /**
+     * @dataProvider type_provider
+     */
+    public function test_backup_restore($type) {
```
Comment (andrewnicols) [inline tests/backup_restore_test.php:65]: Please define all return types.
This one should be :void

File: tests/backup_restore_test.php [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        global $DB;
+        $this->resetAfterTest();
+        $bbactivity = $this->generator->create_module(
+            'bigbluebuttonbn',
+            array('course' => $this->get_course()->id, 'type' => $type),
```
Comment (andrewnicols) [inline tests/backup_restore_test.php:70]: Moodle coding style guidelines encourage [] array syntax for all new code. Please don't use array() unless it is in existing files and to meet existing coding style.

File: tests/backup_restore_test.php [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+     */
+    public function test_backup_restore_with_recordings($type, $nbrecordings) {
+        global $DB;
+        $this->resetAfterTest();
+        // This is for imported recording.
+        $othercourse = $this->generator->create_course();
+        $bbactivityothercourse = $this->generator->create_module(
+            'bigbluebuttonbn',
+            array('course' => $this->get_course()->id, 'type' => instance::TYPE_ALL),
+            ['visible' => true]);
```
Comment (andrewnicols) [inline tests/backup_restore_test.php:122]: Coding style dictates that the ); be on it's own line.

File: tests/backup_restore_test.php [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        // Then import the recordings into the instance.
+        $instance = instance::get_from_instanceid($bbactivity->id);
+        foreach ($recordings as $rec) {
+            $rentity = recording::get_record(['id' => $rec->id]);
+            if ($rentity->get('bigbluebuttonbnid') != $instance->get_instance_id()) {
+                $rentity->create_imported_recording($instance);
+            }
+        }
+        $newcourseid = $this->backup_and_restore($this->get_course());
+        $newbbb = $DB->get_record('bigbluebuttonbn', ['course' => $newcourseid, 'name' => 'BBB Activity'], '*',
```
Comment (andrewnicols) [inline tests/backup_restore_test.php:160]: Should ideally be split onto multiple lines. See coding style guidelines. We are moving towards PSR-12

File: classes/test/testcase_helper_trait.php [hunk 7]
```diff
@@ -80,17 +82,25 @@ protected function get_form_data_from_instance($bbactivity, $course = null) {
...
+            $this->course = $this->generator->create_course(['enablecompletion' => 1]);
+        }
+        return $this->course;
+    }
     /**
      * Setup
      *
      * Enable completion and create a course
      */
-    public function setUp(): void {
```
Comment (andrewnicols) [inline classes/test/testcase_helper_trait.php:88]: Why rename? Should be fine to use `setUp()` still

File: tests/local/helpers/roles_test.php [hunk 8]
```diff
@@ -37,7 +37,15 @@
  * @covers \mod_bigbluebuttonbn\local\helpers\roles
  * @coversDefaultClass \mod_bigbluebuttonbn\local\helpers\roles
  */
-class roles_test extends testcase_helper {
+class roles_test extends \advanced_testcase {
+    use testcase_helper_trait;
+    /**
+     * Setup basic
+     */
+    public function setUp(): void {
```
Comment (andrewnicols) [inline tests/local/helpers/roles_test.php:45]: This change shoulnd't be necessary. no need to rename setUp to basic_setup in the trait.

**Review by andrewnicols — CHANGES_REQUESTED** [review]
Some unnecessary and confusing changes here.

Ideally we should be fixing the testcase_helper trait to not perform setup all the time anyway, and just have it provide data as a return, i.e.

```
protected function get_instance(array $params = []): array {
    $course = ...
    $record = $this->getDataGenerator()->create_module(...);

    return [
        'course' => $course,
        'record' => $record,
        'instance' => instance::get_from_instanceid($record->id),
    ];
}
```

File: classes/test/testcase_helper_trait.php (hunk 1 above)
Comment (laurentdavid) [reply classes/test/testcase_helper_trait.php:20]: That was in the PHP Docs comment, added by mistake.

File: tests/backup_restore_test.php (hunk 3 above)
Comment (laurentdavid) [reply tests/backup_restore_test.php:65]: Yes, I often leave it out, I need to adjust to the new convention here.

File: tests/backup_restore_test.php (hunk 4 above)
Comment (laurentdavid) [reply tests/backup_restore_test.php:70]: Ok

File: classes/test/testcase_helper_trait.php (hunk 2 above)
Comment (laurentdavid) [reply classes/test/testcase_helper_trait.php:161]: I retract my statement :)
It seems that in PHP storm, if the trait is used somewhere in a class that inherit from the right methods, it will not show an error. That's strange, but if this is true, that's quite clever...

File: tests/local/helpers/roles_test.php (hunk 8 above)
Comment (laurentdavid) [reply tests/local/helpers/roles_test.php:45]: I went into a different direction: the setup was just declaring the generator, which in most case is used once, or if more can also be a local variable. I removed the setup completely. If it was more complex then we might need to use a method like you stated in your first comment. I will open another pull request.

**Comment by laurentdavid** [comment]
Andrew, could you check #450 to see if it is better ?
